### PR TITLE
Adding Timer32 and refactoring timer module

### PR DESCRIPTION
**Pull request by JoseClaudioSJr**

### CODE REVIEW

File: src/timer.rs [hunk 1]
```diff
@@ -146,125 +179,149 @@ macro_rules! timer {
...
                 #[inline]
                 fn clear_timer(&mut self) {
-                    self.set_ctl(0x01, 0x02);
+                    self.tim.tax_ctl.modify(|_, w| w.taclr().set_bit());
                     self.count = 0;
                 }
 
                 #[inline]
                 fn setup_timer(&self, count: Count) -> bool {
+                    use TimerUnit::*;
```
Comment (caiotavares) [inline src/timer.rs]: Maybe extract to the top of the file with the other imports

File: src/timer.rs [hunk 2]
```diff
@@ -146,125 +179,149 @@ macro_rules! timer {
...
                 }
 
-                fn set_clock_source(&self, value: u8, prescaler: u8) {
-                    self.clear_ctl(0x0F, 0x06);
-                    self.set_ctl((value | prescaler) as u16, 0x06);
-                }
+                #[inline]
+                fn setup_count(&self, count: Count) -> bool {
+                    use TimerUnit::*;
+                    use ClockSource::*;
```
Comment (caiotavares) [inline src/timer.rs]: Same as line 188

File: src/timer.rs [hunk 3]
```diff
@@ -146,125 +179,149 @@ macro_rules! timer {
...
-                        17..=20 => self.setup_prescaler(4, 2),
-                        21..=24 => self.setup_prescaler(5, 2),
-                        25..=28 => self.setup_prescaler(6, 2),
-                        29..=32 => self.setup_prescaler(7, 2),
-                        33..=40 => self.setup_prescaler(4, 3),
-                        41..=48 => self.setup_prescaler(5, 3),
-                        49..=56 => self.setup_prescaler(6, 3),
-                        _ => self.setup_prescaler(7, 3),
+                fn get_prescaler(&self, min_prescaler: u8) -> ClockSourcePrescaler {
+                    use ClockSourcePrescaler::*;
```
Comment (caiotavares) [inline src/timer.rs]: Same as line 188

File: src/timer.rs [hunk 4]
```diff
@@ -146,125 +179,149 @@ macro_rules! timer {
...
-                    });
-                    [divider, val2, 2u8.pow(val2.into())*(divider+1)]
-                }
-
-                #[inline]
-                fn start_timer(&self) {
-                    self.set_ctl(0x01, 0x04);
-                }
+                fn setup_prescaler(&self, prescaler: ClockSourcePrescaler) -> ClockSourcePrescaler {
+                    use ClockSourcePrescaler::*;
```
Comment (caiotavares) [inline src/timer.rs]: Same as line 188

File: src/timer.rs [hunk 5]
```diff
@@ -146,125 +179,149 @@ macro_rules! timer {
...
+                    if count_ratio != 0 {
+                        min_prescaler = (count_ratio / MAX_COUNT) as u8;
+
+                        // In period, prescaler need to be above the min_prescaler value
+                        if count.1 != Hertz {
+                            min_prescaler = min_prescaler + 1;
+                        }
+
+                        real_prescaler = self.get_prescaler(min_prescaler);
+                        tick_count = (count_ratio / real_prescaler as u32) as u16;
```
Comment (caiotavares) [inline src/timer.rs]: There's no need to declare these vals before actually needing them

```suggestion
                        let real_prescaler = self.get_prescaler(min_prescaler);
                        let tick_count = (count_ratio / real_prescaler as u32) as u16;
```

File: src/timer.rs [hunk 6]
```diff
@@ -146,125 +179,149 @@ macro_rules! timer {
...
+                    let smclk_ratio : u32;
+                    let tick_count : u16;
+                    let mut count_ratio : u32;
+                    let mut clock_source : ClockSource;
+                    let mut is_config: bool;
+                    let mut min_prescaler : u8;
+
+                    is_config = false;
+                    count_ratio = 0;
+                    clock_source = Smclk;
```
Comment (caiotavares) [inline src/timer.rs]: What does `is_config` means in this fn context?

```suggestion
                    let aclk_ratio: u32;
                    let smclk_ratio : u32;
                    let mut count_ratio : u32  = 0;
                    let mut clock_source : ClockSource = Smclk;
                    let mut is_config: bool = false;
```

File: src/timer.rs [hunk 7]
```diff
@@ -146,125 +179,149 @@ macro_rules! timer {
...
-                    } else if (frequency > (self.clocks.aclk.0 / max_period)) {
-                        let prescaler = self.get_prescaler((self.clocks.aclk.0/(MAX_COUNT * frequency)) as u8);
-                        let count = (self.clocks.aclk.0 / (frequency * prescaler[2] as u32)) as u16;
-                        self.set_clock_source(0x04, prescaler[1]);
-                        self.setup_count(count);
-                        true
-                    } else {
-                        false
+                    if count_ratio != 0 {
+                        min_prescaler = (count_ratio / MAX_COUNT) as u8;
```
Comment (caiotavares) [inline src/timer.rs]:
```suggestion
                        let min_prescaler = (count_ratio / MAX_COUNT) as u8;
```

File: src/timer.rs [hunk 8]
```diff
@@ -146,125 +179,149 @@ macro_rules! timer {
...
+                        },
+                        _20 | _24 | _28 | _32 => {
+                            self.tim.tax_ctl.modify(|_,w| w.id().id_2());
+                            self.tim.tax_ex0.modify(|_, w| unsafe {w.bits(prescaler as u16 / 4 -1)} );
+                        },
+                        _40 | _48 | _56 | _64 => {
+                            self.tim.tax_ctl.modify(|_,w| w.id().id_3());
+                            self.tim.tax_ex0.modify(|_, w| unsafe {w.bits(prescaler as u16 / 8 -1)} );
+                        },
+                    }
```
Comment (caiotavares) [inline src/timer.rs]: Much better!

File: src/timer.rs (hunk 6 above)
Comment (JoseClaudioSJr) [reply src/timer.rs]: The try_start function return an error (Error::Unreachable) if the counting value is out of the bounds (is_config == false). Return true or false will take the same effect.

File: src/timer.rs (hunk 7 above)
Comment (JoseClaudioSJr) [reply src/timer.rs]: OK

File: src/timer.rs (hunk 2 above)
Comment (JoseClaudioSJr) [reply src/timer.rs]: TimerUnit::Hertz conflicts with crate::time::Hertz.

File: src/timer.rs (hunk 6 above)
Comment (caiotavares) [reply src/timer.rs]: I think it's simpler to just return true at the end of the block and kill this val

**Review by caiotavares — DISMISSED** [review]

File: Cargo.toml [hunk 9]
```diff
@@ -10,14 +10,24 @@ edition = "2018"
...
 nb = "1.0.0"
 
+[profile.dev]
+debug = true
+debug-assertions = true
+opt-level = "s"
+
+[profile.release]
+opt-level = "s"
+lto = true
```
Comment (caiotavares) [inline Cargo.toml:30]: What does this do?

File: src/timer.rs [hunk 10]
```diff
@@ -321,3 +372,369 @@ timer! {
     TIMER_A2, tim2,
     TIMER_A3, tim3
 }
+
+/// ********************************            TIMER 32            ********************************
```
Comment (caiotavares) [inline src/timer.rs]: It might be a good idea to split this into a new file

Comment (JoseClaudioSJr) [reply src/timer.rs]: Indeed

File: Cargo.toml (hunk 9 above)
Comment (JoseClaudioSJr) [reply Cargo.toml:30]: The opt-level setting controls the -C opt-level flag which controls the level of optimization. "s": optimize for binary size.

The lto setting controls the -C lto flag which controls LLVM's link time optimizations. LTO can produce better optimized code, using whole-program analysis, at the cost of longer linking time.true or "fat": Performs "fat" LTO which attempts to perform optimizations across all crates within the dependency graph.

https://doc.rust-lang.org/cargo/reference/profiles.html

**Review by caiotavares — APPROVED** [review]

**Pull request merged**